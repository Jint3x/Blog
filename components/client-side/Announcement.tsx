import React from "react"


interface Announcement_I {
    title: String,
    desc: String,
    bgColor: String,
    textColor: String,
    headerColor: String,
    borderColor: String,
}

function Announcement(props: Announcement_I) {
    return (
        <div className={`${props.borderColor} ${props.bgColor} p-2 border-solid border-l-4`}>
            <h3 className={`${props.headerColor} text-lg md:text-xl`}>
                {props.title}
            </h3>
            <p className={`${props.textColor} text-base pt-2 pb-2`}>
                {props.desc}
            </p>
        </div>
    )
}


export default Announcement