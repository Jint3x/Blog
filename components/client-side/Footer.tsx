import React from "react"


function Footer() {
    return (
        <footer className={"w-full bg-black"}>
            <div className={"pt-4 pb-4 flex w-56 justify-between mr-auto ml-auto sm:w-72"}>
                <div>
                    <h3 className={"text-main-green text-base"}>About me</h3>

                    <FooterLink 
                     href={"https://devjint3x.com"}
                     text={"My portfolio"}
                    />

                    <FooterLink 
                     href={"https://www.linkedin.com/in/alexandar-dimitrov-546782216/"}
                     text={"Linkedin"}
                    />  
                </div>
                <div>
                    <h3 className={"text-main-green text-base"}>Find me</h3>

                    <FooterLink 
                     href={"https://github.com/Jint3x"}
                     text={"Github"}
                    />

                    <FooterLink 
                     href={"https://twitter.com/Jint3x"}
                     text={"Twitter"}
                    />
                </div>
            </div>
        </footer>
    )
}


interface FooterLink_I {
    href: string,
    text: string,
}

function FooterLink(props: FooterLink_I) {
    return (
        <a 
         href={props.href}
         className={"text-main-blue text-sm no-underline pt-1 pb-1 block"}
        >
            {props.text}
        </a>
    )
}


export default Footer