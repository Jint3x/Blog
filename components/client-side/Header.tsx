import React, { useEffect, useState } from "react"


function Header() {
    let [isVisible, setIsVisible]: [any, any] = useState("inital");

    useEffect(() => {
        window.addEventListener("scroll", () => scrollFunction())

        // Used to detect when the page scroll is below 50px, if so, some new styles will be applied to the header.
        function scrollFunction() {
            if ((document.documentElement.scrollTop || document.body.scrollTop) > 75 && isVisible) {
                setIsVisible(false);
            } else if ((document.documentElement.scrollTop || document.body.scrollTop) < 75 && !isVisible) {
                setIsVisible(true)
            }
        }

        return (
            window.removeEventListener("scroll", scrollFunction)
        )
    }, [])

    return (
        <nav>
            <div className={`
            bg-black flex pt-1 pb-2 content-center flex-col fixed w-full transition-all transform duration-1000 z-50
            ${isVisible !== false ? "" : "-translate-y-10 sm:-translate-y-0"} sm:flex-row sm:pt-1 sm:h-10
            ${isVisible === "initial" ? "" : isVisible === true ? "sm:animate-headerResizeOpposite" : "sm:animate-headerResize"}
            `}>
                <LogoText />
                <Navigation />
            </div>
        </nav>
    )
}

function LogoText() {
    return (
        <div className={`w-full text-center sm:w-40 pl-3 pb-3`}>
                <p className={"text-2xl"}>
                    <span className={"text-main-blue relative bottom-2"}>Jint3x</span>
                    <span className={"text-main-green relative left-1"}>Blogs</span>
                </p>
        </div>
    )
}


function Navigation() {
    return (
        <div className={"w-full self-center sm:w-72 sm:pt-1 sm:ml-auto sm:mr-6"}>
                <ul className={"flex justify-around w-full"}>
                    <NavigationLink 
                     href={"/"}
                     text={"Home"}
                    />

                    <NavigationLink 
                     href={"/blogs"}
                     text={"Blogs"}
                    />

                    <NavigationLink 
                     href={"https://devjint3x.com"}
                     text={"About me"}
                    />
                </ul>
        </div>
    )
}

interface NavigationLink_I {
    text: string,
    href: string,
}

function NavigationLink(props: NavigationLink_I) {
    let [url, setURL] = useState("");

    useEffect(() => {
        setURL(window?.location?.pathname || "")
    })

    let normalNavigation = "text-main-blue"
    let selectedNavigation = "text-main-green bg-main-blue"

    return (
        <li className={`border-main-green border rounded w-20 text-center ${url === props.href ? selectedNavigation : normalNavigation} hover:bg-main-blue-lighter transition-all cursor-pointer`}>
            <a
             href={props.href}
             className={"w-20"}
            >
                {props.text}
            </a>
        </li>
    )
}



export default Header