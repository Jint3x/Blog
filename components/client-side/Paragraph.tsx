import React from "react"

interface ParagraphMainSection_I {
    thumbnail: string,
    thumbnailAlt: string,
    title: string, 
    text: string
}

function ParagraphMainSection(props: ParagraphMainSection_I) {
    return (
        <section>
        <div className={"pt-6 pb-4 w-full"}>
            <img 
             src={props.thumbnail}
             alt={props.thumbnailAlt}
             className={"pt-2 pb-2"}
            />

            <h1 className={"w-full color-default-dark pt-2 pb-2 text-2xl md:text-3xl font-medium text-center"}>
                {props.title}
            </h1>

            <ParagraphTextOnly text={props.text} />
        </div>
        </section>
    )
}


interface ParagraphSection_I {
    title: string,
    text: string,
}

function ParagraphSection(props: ParagraphSection_I) {
    return (
        <section>
        <div className={"pt-6 pb-4 w-full"}>
            <h2 className={"w-full pb-2 color-default-dark text-xl md:text-2xl font-medium"}>
                {props.title}
            </h2>

            <ParagraphTextOnly text={props.text} />
        </div>
        </section>
    )
}


interface ParagraphSubSection_I {
    title: string,
    text: string, 
}

function ParagraphSubSection(props: ParagraphSubSection_I) {
    return (
        <div className={"pt-4 pb-2 w-full"}>
            <h3 className={"w-full color-default-dark text-lg md:text-xl font-medium"}>
                {props.title}
            </h3>

            <ParagraphTextOnly text={props.text} />
        </div>
    )
}


interface ParagraphTextOnly_I {
    text: string,
}


function ParagraphTextOnly(props: ParagraphTextOnly_I) {
    return (
        <p className={"w-full pt-1 pb-2 color-default-dark text-base md:text-lg"}>
            {props.text}
        </p>
    )
}


export { ParagraphMainSection, ParagraphSection, ParagraphSubSection, ParagraphTextOnly }