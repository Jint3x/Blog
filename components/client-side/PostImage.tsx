import React from "react"

type ImageType = "wide" | "tall";
interface PostImage_I {
    src: string, 
    alt: string,
    desc: string,
    imageType: ImageType
}

function PostImage(props: PostImage_I) {
    return (
        <div className={"pt-4 pb-4"}>
            <img 
             src={props.src}
             alt={props.alt}
             className={`block ml-auto mr-auto ${props.imageType === "wide" ? "w-full sm:w-128" : "w-4/6 sm:w-72 "}`}
            />

            <p className={`ml-auto mr-auto pt-1 text-sm text-color-image-comment ${props.imageType === "wide" ? "w-full sm:w-128" : "w-4/6 sm:w-72 "}`}>
                {props.desc}
            </p>
        </div>
    )
}


export default PostImage