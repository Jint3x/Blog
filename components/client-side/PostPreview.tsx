import React from "react"


interface PostPreview_I {
    thumbnail: string,
    title: string,
    description: string,
    date: string, 
    tags: string[],
    url: string,

    // If loading is set to true, all other fields won't be used
    loading?: boolean,
}

function PostPreview(props: PostPreview_I) {
    return (
        <div className={"w-72 bg-white ml-auto mr-auto h-80 rounded shadow-lg shadow shadow-md shadow-personal mt-12 mb-12 overflow-hidden"}>

            {
                props.loading ? 
                (
                    <div className={"h-20per bg-black"} />
                )
                :
                (
                    <div 
                     style={{backgroundImage: `url(${props.thumbnail})`}} 
                     className={"h-20per bg-center overflow-hidden cursor-pointer"} 
                     onClick={() => location.assign(props.url)}
                    />
                )
            }
            

            <div className={"h-65per color-default-dark relative"}>
                <h3 className={"text-center pt-1 pb-2 text-2xl overflow-hidden"}>
                    {
                        props.loading ? "Loading..." : props.title
                    }
                </h3>

                <div className={"relative h-3/6 pl-2 pr-2 mt-2"}>
                    <p className={"overflow-hidden h-full"}>
                        {props.loading ? "" : props.description}
                    </p>
                    <div className={"absolute bottom-0 h-12 w-full bg-gradient-to-t from-white to-transparent"} />
                </div>
                

                <p className={"absolute bottom-1 text-sm left-1 text-post-date"}>
                    {props.loading ? "" : props.date}
                </p>
            </div>

            <div 
             className={"h-15per bg-main-blue-lighter flex-nowrap flex items-center overflow-hidden relative cursor-pointer"}
             onClick={() => location.assign(props.url)}
            >
                <div className={"absolute h-full w-1/12 right-0 bg-gradient-to-l from-main-blue-lighter to-transparent"} />
                {
                    props.loading ? "" : props.tags.map(tag => <RenderTag name={tag} key={tag} />)
                }
            </div>
        </div>
    )
}


interface RenderTag_I {
    name: string,
}

function RenderTag(props: RenderTag_I) {
    return (
        <p className={"bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"}>
            {props.name}
        </p>
    )
}




export default PostPreview