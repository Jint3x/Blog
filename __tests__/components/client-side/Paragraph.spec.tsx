/**
 * @jest-environment jsdom
 */
import { ParagraphMainSection, ParagraphSection, ParagraphSubSection, ParagraphTextOnly }  from "../../../components/client-side/Paragraph";
import ReactRenderer from "react-test-renderer";
import { screen, render } from "@testing-library/react";


test("Renders a full introduction component", () => {
    render(
        <ParagraphMainSection
         thumbnail={"/special/image"}
         thumbnailAlt={"Special Image"}
         title={"A test title"}
         text={"A test text"}
        />
    )
    
    let image = screen.getByAltText("Special Image");
    let title = screen.getByText("A test title");
    let text = screen.getByText("A test text");

    
    expect(image.nodeName).toBe("IMG");
    expect(title.nodeName).toBe("H1");
    expect(text.nodeName).toBe("P");
})


test("Renders a section component", () => {
    render(
        <ParagraphSection
         title={"My Section!"}
         text={"This is my section!"}
        />
    )

    let title = screen.getByText("My Section!");
    let text = screen.getByText("This is my section!");
    let levelChildrenCount = title.parentElement?.childElementCount;

    expect(title.nodeName).toBe("H2");
    expect(text.nodeName).toBe("P");
    expect(levelChildrenCount).toBe(2);
})


test("Renders a sub section component", () => {
    render(
        <ParagraphSubSection
         title={"My Sub Section!"}
         text={"This is my sub section!"}
        />
    )

    let title = screen.getByText("My Sub Section!");
    let text = screen.getByText("This is my sub section!");
    let levelChildrenCount = title.parentElement?.childElementCount;

    expect(title.nodeName).toBe("H3");
    expect(text.nodeName).toBe("P");
    expect(levelChildrenCount).toBe(2);
})


test("Renders a section with only text", () => {
    render(
        <ParagraphTextOnly
         text={"This is my sub section!"}
        />
    )

    let text = screen.getByText("This is my sub section!");
    let levelChildrenCount = text.parentElement?.childElementCount;

    expect(text.nodeName).toBe("P");
    expect(levelChildrenCount).toBe(1);
})


// A single test will handle all of the 4 paragraph varients.
test("Take a snapshot of all of the different usages of the component. Make multiple tests", () => {
    let treeMainSection = ReactRenderer.create(
        <ParagraphMainSection
         thumbnail={"/special/image"}
         thumbnailAlt={"Special Image"}
         title={"A test title"}
         text={"A test text"}
        />
    )
    .toJSON();

    let treeSection = ReactRenderer.create(
        <ParagraphSection
         title={"My Section!"}
         text={"This is my section!"}
        />
    ) 
    .toJSON();

    let treeSubSection = ReactRenderer.create(
        <ParagraphSubSection
         title={"My Sub Section!"}
         text={"This is my sub section!"}
        />
    )
    .toJSON();

    let treeTextOnly = ReactRenderer.create(
        <ParagraphTextOnly
         text={"This is my sub section!"}
        />
    )
    .toJSON();

    expect(treeMainSection).toMatchSnapshot();
    expect(treeSection).toMatchSnapshot();
    expect(treeSubSection).toMatchSnapshot();
    expect(treeTextOnly).toMatchSnapshot();
})