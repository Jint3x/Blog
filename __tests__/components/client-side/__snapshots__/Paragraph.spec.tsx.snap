// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Take a snapshot of all of the different usages of the component. Make multiple tests 1`] = `
<div
  className="pt-6 pb-4 w-full"
>
  <img
    alt="Special Image"
    className="pt-2 pb-2"
    src="/special/image"
  />
  <h1
    className="w-full color-default-dark pt-2 pb-2 text-2xl md:text-3xl font-medium text-center"
  >
    A test title
  </h1>
  <p
    className="w-full pt-1 pb-2 color-default-dark text-base md:text-lg"
  >
    A test text
  </p>
</div>
`;

exports[`Take a snapshot of all of the different usages of the component. Make multiple tests 2`] = `
<div
  className="pt-6 pb-4 w-full"
>
  <h2
    className="w-full pb-2 color-default-dark text-xl md:text-2xl font-medium"
  >
    My Section!
  </h2>
  <p
    className="w-full pt-1 pb-2 color-default-dark text-base md:text-lg"
  >
    This is my section!
  </p>
</div>
`;

exports[`Take a snapshot of all of the different usages of the component. Make multiple tests 3`] = `
<div
  className="pt-4 pb-2 w-full"
>
  <h3
    className="w-full color-default-dark text-lg md:text-xl font-medium"
  >
    My Sub Section!
  </h3>
  <p
    className="w-full pt-1 pb-2 color-default-dark text-base md:text-lg"
  >
    This is my sub section!
  </p>
</div>
`;

exports[`Take a snapshot of all of the different usages of the component. Make multiple tests 4`] = `
<p
  className="w-full pt-1 pb-2 color-default-dark text-base md:text-lg"
>
  This is my sub section!
</p>
`;
